bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HostnameKeyStoreMapping.java

```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.io.net.http;
+
+import java.security.KeyStore;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * Provides a TrustManager for supported endpoints
+ *
+ * @author Martin van Wingerden - Initial Contribution
+ */
+@NonNullByDefault
+public interface HostnameKeyStoreMapping {
+    /**
+     * hostname for which this TrustManager is intended
+     *
+     * @return
+     */
+    String getHostName();
+
+    /**
+     * A KeyStore for the specified hostname
+     *
+     * @return
+     */
+    KeyStore getKeyStore();
```

Comment:
Why not let the provider directly provide certificate resources and have them wrapped into KeyStores by the framework? I am pretty sure that many people would not know, how to provide a KeyStore here.

Reply:
This is a point where we were still in doubt in the related issue, would it be enough to only have an option to provide a certificate. Or should there also be an option to provide a TrustManager so that you can really tweak everything? 

I do indeed agree that the abstraction of a keystore is not really useful it was mainly chosen for technical reasons.

Note however that to accept all certificates for a single host or similar requirements something like providing a TrustManager is needed..

Or we could make that an option of the HttpUtil because a lot of self hosted services run on local host(s) so working on the common name / SAN could be a problem anyway.

Reply:
Hm, couldn't there different interfaces be supported?
For example (which one needs to be discusses):
* `HostnameCeritifcateMapping`
* `HostnameKeyStoreMapping`
* `HostnameTrustStoreMapping`

If a lookup is done, we could define that the most customizable wins. First check if there is a truststore, then if there is a keystore, then if there is a certificate mapping...

Reply:
Nice idea!

Only mildly complicating thing is that we have both the common name and the subject alternative names, so it might become an intensive lookup for every request. Because some certificates have 100 SANs so that would be 300 lookups if none is available. But definitely worth discussing and exploring this.

Reply:
They would still register for specific endpoints and in practice, I would think that there'll always only be one service registered per domain name.

I agree that having different interfaces for different complexities make sense - and for the start, I think the certificate provider is all we need for now (anything else can be added later).

Reply:
> I think the certificate provider is all we need for now

Which one do you refer by "we"?
Is this (certificate mapping) what is needed by ESH bindings?

AFAIK (need to inspect the code next week) but "me" (or "we" as a company that is using the ESH framework) rely on customized trustmanagers.

Reply:
I’ll create something which covers those two scenarios then 👍

Reply:
> Which one do you refer by "we"?

To us as the framework developers&contributors to fulfill a currently existing requirement that caused this PR to be created (https://github.com/openhab/openhab2-addons/issues/3762).

> but "me" rely on customized trustmanagers.

So you already have such a feature build for yourself? How are you using it without the framework offering such extension points?

Reply:
Note that the `TrustManagerProvider` returns a `TrustManager` given an `endpoint`. 

If we want to replace the `TrustManagerProvider` it would make sense because it allows for an easy migration from the `TrustManagerProvider` to the new implementation.

Reply:
> So you already have such a feature build for yourself? How are you using it without the framework offering such extension points?

How have bindings used HTTP clients before the ESH framework provides support for them?
They used Apache HTTP client of Jetty implementation without framework support.

Reply:
> Note that the TrustManagerProvider returns a TrustManager given an endpoint.

Which should be deprecated and kept as it is - so I would suggest to use a different interface name here.

Reply:
> They used Apache HTTP client of Jetty implementation without framework support.

@maggu2810 So you're saying that you would like to use the shared Jetty instance with that trust extension mechanism? Then sure, besides the certificate provider, we can also directly add an interface for custom trust managers, if this is what you'd additionally required. My point was only that there isn't any need to implement a feature that nobody has asked for yet.